src/wh_server.c

```diff
@@ -28,8 +30,613 @@
+    XMEMCPY((uint8_t*)server->cache[foundIndex].meta, (uint8_t*)meta,
+        sizeof(NvmMetaData));
+#ifdef WOLFHSM_SHE_EXTENSION
+    /* if this was RAM_KEY, set the global so we can export */
+    if (meta->id == WOLFHSM_SHE_TRANSLATE_KEY_ID(WOLFHSM_SHE_RAM_KEY_ID))
+        hsmSheRamKeyPlain = 1;
+#endif
+    return 0;
+}
+
+static int hsmReadKey(whServerContext* server, NvmMetaData* meta, uint8_t* out)
+{
+    int i;
+    int ret = 0;
+    uint16_t searchId = meta->id;
+    uint16_t outLen = meta->len;
+    /* make sure id is valid */
+    if (meta->id == 0)
+        return BAD_FUNC_ARG;
+    /* check the cache */
+    for (i = 0; i < WOLFHSM_NUM_RAMKEYS; i++) {
+        /* copy the meta and key before returning */
+        if (server->cache[i].meta->id == searchId) {
+            /* check outLen */
+            if (server->cache[i].meta->len > outLen)
+                return BUFFER_E;
+            XMEMCPY((uint8_t*)meta, (uint8_t*)server->cache[i].meta,
+                sizeof(NvmMetaData));
+            XMEMCPY(out, server->cache[i].buffer, meta->len);
+            return 0;
+        }
+    }
+#if 0
+    /* setup address */
+    if (ctx->partition == 1)
+        addr = WOLFHSM_PART_1;
+    /* find id in list */
+    for (i = 0; i < WOLFHSM_KEYSLOT_COUNT; i++) {
+        /* check id */
+        if (ctx->nvmMetaCache[i].id == searchId)
+            break;
+        /* keep track of length used by previous keys */
+        keyOffset += ctx->nvmMetaCache[i].len;
+    }
+    /* if we looped to the end, no match */
+    if (i >= WOLFHSM_KEYSLOT_COUNT)
+        ret = BAD_FUNC_ARG;
+    /* check the length, meta->len is the return size */
+    if (ret == 0 && ctx->nvmMetaCache[i].len > outLen)
+        ret = BUFFER_E;
+    /* read key */
+    if (ret == 0) {
+        ret = hal_flash_read(addr + WOLFHSM_HEADER_SIZE + keyOffset, out,
+            ctx->nvmMetaCache[i].len);
+    }
+    /* calculate key digest */
+    if (ret == 0)
+        ret = wc_InitSha256_ex(sha, ctx->heap, ctx->devId);
+    if (ret == 0)
+        ret = wc_Sha256Update(sha, out, ctx->nvmMetaCache[i].len);
+    if (ret == 0)
+        ret = wc_Sha256Final(sha, digest);
+    /* validate key */
+    if (XMEMCMP(digest, ctx->nvmMetaCache[i].confDigest, WOLFHSM_DIGEST_STUB)
+        != 0) {
+        ret = BAD_FUNC_ARG;
+    }
+    if (ret == 0) {
+        /* cache key if free slot, will only kick out other commited keys */
+        hsmCacheKey(ctx, &ctx->nvmMetaCache[i], out);
+        /* update out meta with nvm cache contents */
+        XMEMCPY(meta, &ctx->nvmMetaCache[i], sizeof(NvmMetaData));
+    }
+#ifdef WOLFHSM_SHE_EXTENSION
+    /* use empty string if we couldn't find the master ecu key */
+    if (ret != 0 && searchId == WOLFHSM_SHE_MASTER_ECU_KEY_ID) {
+        XMEMSET(out, 0, WOLFHSM_SHE_KEY_SZ);
+        meta->len = WOLFHSM_SHE_KEY_SZ;
+        meta->id = searchId;
+        ret = 0;
+    }
+#endif
+#endif
+    return ret;
```

Comment:
is it possible to not get a hit on the key in the cache? It looks like this unconditionally returns success.

Reply:
the dead code handled the miss case, updated since nvm isn't working